src/history/hs_rec.c

```diff
@@ -511,8 +511,12 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_MULTI *mult
              * No need to continue if we found a first self contained value that is globally
              * visible.
              */
-            if (__wt_txn_upd_visible_all(session, upd) && WT_UPDATE_DATA_VALUE(upd))
+            if (__wt_txn_upd_visible_all(session, upd) && WT_UPDATE_DATA_VALUE(upd)) {
+                /* Free obsolete updates if exist. */
+                if (!F_ISSET(r, WT_REC_EVICT) && upd->next != NULL)
+                    __wt_update_obsolete_check(session, r->ref, upd, false);
                 break;
```

Comment:
Why we don't need lock here?

Reply:
As part of the reconciliation, the `page lock` is already acquired, so there is no need to acquire it again.